srk/src/numberField.ml

```diff
@@ -0,0 +1,994 @@
+    end
+  module O : sig
+    type ideal
+    type o
+    val pp_o : Format.formatter -> o -> unit
+    val make_o_el : elem -> ZZ.t * o
+    val order_el_to_f_elem : ZZ.t * o -> elem
+    val idealify : ZZ.t array array -> ideal
+    val pp_i : Format.formatter -> ideal -> unit
+    val equal_i : ideal -> ideal -> bool
+    val sum_i : ideal -> ideal -> ideal
+    val mul_i : ideal -> ideal -> ideal
+    val ideal_generated_by : o -> ideal
+    val one_i : ideal
+    val intersect_i : ideal -> ideal -> ideal
+    val quotient_i : ideal -> ideal -> ideal
+    val get_smallest_int : ideal -> ZZ.t
+    type frac_ideal
+    val one : frac_ideal
+    val sum : frac_ideal -> frac_ideal -> frac_ideal
+    val intersect : frac_ideal -> frac_ideal -> frac_ideal
+    val mul : frac_ideal -> frac_ideal -> frac_ideal
+    val exp : frac_ideal -> int -> frac_ideal
+    val quotient : frac_ideal -> frac_ideal -> frac_ideal
+    val pp : Format.formatter -> frac_ideal -> unit
+    val equal : frac_ideal -> frac_ideal -> bool
+    val subset : frac_ideal -> frac_ideal -> bool
+    val make_frac_ideal : ZZ.t -> ideal -> frac_ideal
+    val compute_overorder : frac_ideal -> frac_ideal -> frac_ideal * frac_ideal
+    val factor_refinement : frac_ideal list -> frac_ideal * (frac_ideal * int) list
+    val compute_factorization : frac_ideal list -> int list list * frac_ideal list * frac_ideal
+    val find_unit_basis : o list -> int list list
+  end
+  val find_relations : elem list -> int list list
+end
+
+
+
+
+
+
+module MakeNF (A : sig val min_poly : QQX.t end) = struct
+
+  type elem = QQX.t (* elements of the field are polynomials of degree < the degree of the field*)
+
+  let compute_min_poly_p el p = 
+    let open Linear in
+    let p_deg = QQX.order p in
+    if p_deg = 0 then el
+    else
+      let m = Array.make_matrix p_deg (p_deg+1) QQ.zero in (*m is a matrix whose columns consist of the powers of el*)
+      for i = 0 to p_deg do
+        let r = snd (QQX.qr (QQX.exp el i) p) in
+        BatEnum.iter (
+          fun (c, d) -> 
+            m.(d).(i) <- c
+        ) (QQX.enum r)
+      done;
+      (*In general I don't think null is guarenteed to be reduced with respect to powers of el. However,
+         I think it is do to how nullspace is implemented. *)
+      let null = nullspace (QQMatrix.of_dense m) (List.init (p_deg + 1) (fun i -> i)) in
+      let min = snd (List.fold_left (
+        fun (min_deg, min_p) vec ->
+          let poly = QQX.of_enum (QQVector.enum vec) in
+          if QQX.order poly < min_deg then (QQX.order poly, poly)
+          else (min_deg, min_p)
+      ) (p_deg + 2, QQX.one) null) in
+      let lc = QQX.coeff (QQX.order min) min in
+      QQX.map (fun _ c -> QQ.div c lc) min
+
+
+  (* The lcm of the denominators of min_poly*)
+  let min_poly_den_lcm = 
+    QQX.fold (
+      fun _ c l ->
+        ZZ.lcm (QQ.denominator c) l
+    ) A.min_poly (QQ.numerator (QQX.coeff (QQX.order A.min_poly) A.min_poly))
```

Comment:
No, A.min_poly is the polynomial given to construct the NF. It is not assumed to be monic for the implementation. On instatiation of the functor, the implementation computes a monic integer polynomial accessible as int_poly such that A.min_poly and int_poly generate the same number field. int_poly is the the minimal polynomial that will be used for the rest of the calculations.

Reply:
Why is the numerator of the leading coefficient part of the lcm of the denominators of min_poly?

Reply:
If A.min_poly = 4x^2 + 1, the field is Q(i/2). But i/2 is not an algebraic integer. We need the lcm of the monic polynomial that defines Q(i/2). I.e. x^2 + 1/4. The lcm of the dens of x^2 + 1/4 is 4, which is the lc of A.min_poly.

Reply:
I can change the name, or we can factor the numerators of the input polynomials. It doesn't functionally matter either way. The real purpose of min_poly_den_lcm is to recognize the isomorphism between Q[x]/A.min_poly and Q[x]/int_poly. With int_poly a monic integer polynomial. I.e. defines an algebraic integer. There are many choices for int_poly that work, but we just need one. Take A.min_poly = 4x^2 + 2 then Q[x]/A.min_poly = Q(i/sqrt(2)). Taking min_poly_den_lcm = 4 will yield int_poly = x^2 + 8. So Q[x]/int_poly = Q(2isqrt(2)). Q(i/sqrt(2)) = Q(2isqrt(2)) because min_poly_den_lcm * i/sqrt(2) = 2isqrt(2). On the other hand if we want min_poly_den_lcm = 2, then int_poly = x^2 + 2. In this case Q[x]/int_poly = Q(isqrt(2)) and 2*isqrt(2) = 2isqrt(2). Everything is the same field, but just written in different bases. It doesn't matter which base we use so long as the "root" is an algebraic integer. Being an algebraic integer is only important for the order calculations later. So the point is there is no assumption on the factors of the coefficients. We could canonicalize by dividing by the lc, but as stated it doesn't matter for any calculations. In any case I think you'd just like a more descriptive name. But maybe you think we should canonicalize too. It doesn't really matter to me.

Reply:
I meant denominators not numerators in the first sentence.

Reply:
Either way is fine I think -- I just want to avoid revisiting the code in a few years and seeing that the code doesn't *exactly* match the name of the function.

Reply:
I've changed the name